File: Queue/Queue.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define Max_Size 7
typedef struct queue{
	char arr[Max_Size];
	int head;
	int tail;
	int len;
}Queue;

void init(Queue *queue){
	queue->head=-1;
	queue->tail=-1;
}

int is_full(Queue *queue){
	if(queue->tail==Max_Size-1)
		return 1;
	else
		return 0;
}

int is_empty(Queue *queue){
	if(queue->head==queue->tail)
		return 1;
	else
		return 0;
}

int check_blank(Queue *queue){
	if(is_full(queue)&&queue->head!=-1){
	return 1;
	}
	else
		return 0;
}

void shift(Queue *queue){
	int i;
	for(i=0;i<strlen(queue->arr);i++){
		queue->arr[i]=queue->arr[queue->head+1];
		(queue->head)+=1;
	}
	queue->head=-1;
	queue->tail=strlen(queue->arr)-1;
}
void print(Queue *queue){
	if(is_empty(queue)){
		printf("Queue is empty\n");
	}
	else{
  int i;
  for(i=(queue->head)+1;i<=queue->tail;i++){
    printf("%c ",queue->arr[i]);}
  printf("\n");}
}

void enqueue(Queue *queue, char input[]){
	int i;
	for(i=0;i<strlen(input);i++){
	if(input[i]=='+'){
	if(is_full(queue)&&is_empty(queue)){
		init(queue);
		(queue->tail)+=1;
		queue->arr[queue->tail]=input[i+1];
	}
	else if(is_full(queue)){
		if(check_blank(queue)){
					shift(queue);
					(queue->tail)+=1;
					queue->arr[queue->tail]=input[i+1];
				}
		else
			printf("Queue is full\n");
			return;
	}
	else{
		(queue->tail)+=1;
		queue->arr[queue->tail]=input[i+1];
	}
	}
	}
}

void peek(Queue *queue){
	printf("%c\n",queue->arr[queue->head+1]);
}

void dequeue(Queue *queue, char input[]){
	int i;
	for(i=0;i<strlen(input);i++){
		if(input[i]=='-'){
		 printf("return %c\n",queue->arr[queue->head+1]);
		 (queue->head)+=1;	
		}
	}
}

void cdequeue(Queue *queue){
	
	printf("return %c\n",queue->arr[queue->head+1]);
	(queue->head)+=1;
}
void data_count(Queue *queue){
	printf("%d\n",queue->tail-queue->head);
}

void head(Queue *queue){
	printf("%d\n",queue->head+2);
}

void tail(Queue *queue){
	printf("%d\n",queue->tail+1);
}

void is_member(Queue *queue, char input[]){
	int i=queue->head+1;
	int len=queue->tail-queue->head;
	int n=0;
	while(i<=queue->tail){
		if(queue->arr[i]==input[1]){
			break;
		}
		else{
			n+=1;
		}
		i+=1;
	}
	if(n==len){
		printf("FALSE\n");
	}
	else{
		printf("TRUE\n");
	}
}

void replace(Queue *queue, char input[]){
	queue->arr[queue->tail]=input[1];
}

void upper(Queue *queue){
	queue->arr[queue->tail]-=32;
}

void lower(Queue *queue){
	queue->arr[queue->tail]+=32;
}

void menu(){
	printf("----------------stack Menu----------------\n");
	printf("+(data): enqueue\n");
	printf("L: print queue\n");
	printf("P: peek data\n");
	printf("-: dequeue one data\n");
	printf("F: is full?\n");
	printf("#: data count\n");
	printf("T: tail\n");
	printf("H: head\n");
	printf("(number)-: dequeue number of data\n");
	printf("?(data): is member?\n");
	printf("=(data): replace data\n");
	printf("C: clear queue\n");
	printf("M: veiw menu\n");
	printf("U: upper case\n");
	printf("l: lower case\n");
	printf("Q: Quit\n");
	printf("E: is empty?\n");
}

int main(){
	menu();
	Queue *queue= (Queue*)malloc(sizeof(queue));
	init(queue);
	while(1){
		char input[10];
		gets(input);
		if(input[0]=='+'){
			if(is_full(queue)){
				if(check_blank(queue)){
					shift(queue);
					enqueue(queue,input);
					print(queue);
				}
				else
					printf("Queue is full\n");
				}
			else{
				enqueue(queue,input);
				print(queue);
				}
			}
		else if(input[0]=='P'){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				peek(queue);
			}
		}
		else if(input[0]=='-'){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				dequeue(queue,input);
				print(queue);}
			}
		else if(input[0]=='L'){
				print(queue);
			}
		else if(input[0]=='F'){
			if(is_full(queue)){
			if(check_blank(queue)){
				printf("False\n");
			}
			else{
				printf("True\n");
			}
			}
			else{
				printf("False\n");
			}
		}
		else if(input[0]=='E'){
			if(is_empty(queue)){
				printf("True\n");
			}
			else{
				printf("False\n");
			}
		}
		else if(input[0]=='#'){
			data_count(queue);
		}
		else if(input[0]=='H'){
			if(is_empty(queue)){
				printf("-1");
			}
			else{
				head(queue);
			}
		}
		else if(input[0]=='T'){
			if(is_empty(queue)){
				printf("-1");
			}
			else{
				tail(queue);
			}
		}
		else if(input[0]=='?'){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				is_member(queue,input);
			}
		}
		else if(input[0]=='='){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				replace(queue,input);
				print(queue);
			}
		}
		else if(input[0]=='C'){
			init(queue);
			printf("Queue is empty\n");
		}
		else if(input[0]=='U'){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				upper(queue);
				print(queue);
			}
		}
		else if(input[0]=='l'){
			if(is_empty(queue)){
				printf("Queue is empty\n");
			}
			else{
				lower(queue);
				print(queue);
			}
		}
		else if(input[0]=='Q'){
			break;
		}
		else if(input[0]=='M'){
			menu();
		}
		else{
			if(input[1]=='-'){
				int n,i;
				n=atoi(&input[0]);
				for(i=0;i<n;i++){
					if(is_empty(queue)){
						printf("Queue is empty");
						break;
					}
					else{
						cdequeue(queue);
					}
				}
				if(is_empty(queue)){
					printf("\n");
				}
				else{
					print(queue);
				}
			}
		}
		}
}

File: Queue/Queue.py
#shift를 구현할 때 파이썬은 리스트 안의 요소 이동이 불가능하기 때문에
#연결리스트를 사용하여 채워지지 않은 리스트를 사라지게 했습니다.
maxsize=7 #연결리스트이기 때문에 무한하다는 점을 보완하기 위해 최대 크기를 정의하여 유한하게 했습니다,
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Queue:
    def __init__(self):
        self.head = None
        self.tail = None
        self.count = 0

    def is_empty(self):
        if self.count == 0:
            return True
        else:
            return False
    
    def is_full(self):
      global maxsize
      if maxsize==self.count:
        return True
      else:
        return False

    def enqueue(self, item):
        newnode = Node(item)
        if self.is_empty():
            self.head = self.tail = newnode
        else:
            self.tail.next = newnode
            self.tail = newnode
        self.count += 1

    def dequeue(self):
        if self.is_empty():
            print("Queue is EMPTY")
            return
        item = self.head.data
        self.head = self.head.next
        self.count -= 1
        return item

    def peek(self):
        if self.is_empty():
            print("Queue is EMPTY")
            return
        return self.head.data

    def print_queue(self):
        if self.is_empty():
            print("Queue is EMPTY")
            return
        temp = Queue()
        while not self.is_empty():
            print(self.peek(), end = ' ')
            temp.enqueue(self.dequeue())
        while not temp.is_empty():
            self.enqueue(temp.dequeue())
        print()

    def replace(self, item):
        if self.is_empty():
            print("Queue is EMPTY")
            return
        self.tail.data = item

    def is_member(self, item):
        tf = False
        temp = Queue()
        while not self.is_empty():
            if self.peek() == item:
                tf = True
            temp.enqueue(self.dequeue())
        while not temp.is_empty():
            self.enqueue(temp.dequeue())
        return tf

    def clear_queue(self):
        self.head = None
        self.tail = None
        self.count = 0

    def Head(self):
      return self.head.data #연결리스트를 사용했기 때문에 head는 무조건 맨앞이라 인덱스값 대신 요소를 반환했습니다.

    def Tail(self):
      return self.tail.data #head와 동일한 이유

    def upper(self):
        self.tail.data=self.tail.data.upper()

    def lower(self):
        self.tail.data=self.tail.data.lower()
q = Queue()
def menu():
    print("----------------Queue Menu----------------\n");
    print("+(data): enqueue\n");
    print("L: print queue\n");
    print("P: peek data\n");
    print("-: dequeue one data\n");
    print("#: data count\n");
    print("T: tail\n");
    print("H: head\n");
    print("(number)-: dequeue number of data\n");
    print("?(data): is member?\n");
    print("=(data): replace data\n");
    print("C: clear queue\n");
    print("M: veiw menu\n");
    print("U: upper case\n");
    print("l: lower case\n");
    print("Q: Quit\n");
    print("E: is empty?\n");

menu()
while True:
    m = input("")
    i = 0
    while i < len(m):
        if m[i] == '+':
          if q.is_full():
            print("Queue is full")
          else:
            q.enqueue(m[i+1])
            i += 1
        elif m[i] == '-':
            q.dequeue()
        elif m[i] == 'P':
            print("Return %c" %q.peek())
        elif m[i] == '#':
            print("%d" %q.count)
        elif m[i] == 'E':
            if q.is_empty():
                print("TRUE")
            else:
                print("FALSE")
        elif m[i] == '=':
            q.replace(m[i+1])
            i += 1
        elif m[i] == '?':
            if q.is_empty():
                print("Queue is EMPTY")
            elif q.is_member(m[i+1]):
                print("TRUE")
            else:
                print("FALSE")
            i += 1
        elif m[i] == 'C':
            q.clear_queue()
        elif m[i] == 'L':
            q.print_queue()
        elif m[i]=='F':
          if q.is_full():
              print("True")
          else:
            print("False")
        elif m[i]=='H':
          if q.is_empty():
            print("Queue is empty")
          else:
            print("%d"%q.Head())
        elif m[i]=='T':
          if q.is_empty():
            print("Queue is empty")
          else:
            print("%d" %q.Tail())
        elif m[i]=='U':
            if q.is_empty():
                print("Queue is empty")
            else:
                q.upper()
        elif m[i]=='l':
            if q.is_empty():
                print("Queue is empty")
            else:
                q.lower()
        if m[0] == 'Q':
          break
        elif m[0].isdigit():
          n=int(m[0])
          if m[1] == '-':
            for i in range(n):
              q.dequeue()
        i += 1